### ISSUE TITLE
Question: nesting state?

### ISSUE THREAD

**Issue by NathanWalker** [simpulton/ngrx-rest-app#8]
@simpulton Great article, learned a lot from it, thank you!
Just curious, how would you recommend handling this case:

Say you wanted more than 1 observable state property in a service nested under 1 `store` key:
```
@Injectable()
export class LargeService {
  items: Observable<Array<Item>>;
  machines: Observable<Array<Item>>;

  constructor(private http: Http, private store: Store<AppStore>) {
    // this wouldn't work: this.items = store.select('items');
    store.select('largeService').subscribe((state) => {
      this.items = state.items;
      this.machines = state.machines;
    });
  }

....
bootstrap(App, [
  LargeService,
  HTTP_PROVIDERS,
  provideStore({largeService})
])
```

Would that be the best way to handle nested state with `@ngrx`?

**Comment by akaztp** [comment]
I'm working around the same case has you, having a large state with nested and inter-dependent sub-modules.

For your example a simple solution would be:
`this.items = store.select('largeService').map((state) => state.items).distinctUntilChanged();`

But the problem here is that from below the global state there is no push changes detection.
Note that when `state.machines` changes, the above `map()` and `distinctUntilChanged()` also runs making a "dirty checking". And this conflicts with the "redux" paradigm, I think.

**Comment by simpulton** [comment]
You are going to spoil my next blog post but here is the approach I have been using.

1. Normalize your data. Dan Abramov talks about this here https://github.com/gaearon/normalizr

2. Use Observer.combineLatest to create a computed observable within a service. You can see an example of this here https://jsbin.com/gurupe/edit?html,js,output

3. Within ngrx, it would look something like this...
```
@Injectable()
export class ContextService {
  context: Observable<Context>;
  sessions: Observable<Array<Session>>;
  selectedUser: Observable<User>;

  constructor(private store: Store<AppStore>) {
    this.context = Observable.combineLatest(
        store.select('sessions'),
        store.select('selectedUser'),
        (sessions: Session[], selectedUser: User) => {
          return {
            sessions: sessions.filter(session => selectedUser.sessions.indexOf(session.id) > -1)
          };
        });

    this.context
      .subscribe(c => console.log('ContextService.context', c));
  }
};```

And that way, when either of the "tables" change... it is automatically pushed to the view. 

Let me know if I need to elaborate further. Hope this helps!

**Comment by akaztp** [comment]
Yes, I was reading about this way of state dependence today on "[Computing derived data](http://redux.js.org/docs/recipes/ComputingDerivedData.html)".
The bottom line is that if you have some sub-states that depend on each other, take the "computed" one from the state and compute it from the others, like the example above.

The thing is when one needs to use this computed data in several places of the application. I think one solution is to place the service that makes that data available in the hierarchy of providers associated with the DOM components in Angular2. I'm using this approach to create a tree of state "handlers" that act as a proxy and encapsulator to the top level ngrx/state.
Regards
